### ISSUE TITLE
gvapython on a custom classification model

### ISSUE THREAD

**Issue by hungtrieu07** [dlstreamer/dlstreamer#459]
Hi everyone, I'm working on a person attribute recognition model. This is my pipeline:
```
GST_DEBUG=4 gst-launch-1.0 filesrc location=/home/hungtrieu07/dev/video_data/1740731094984.mp4 ! decodebin ! videoconvert ! \
    gvadetect model=ov_exported_model/person_vehicle/FP16.xml device=CPU ! \
    gvawatermark ! \
    gvaclassify model=ov_exported_model/person-attr/model.xml model-proc=ov_exported_model/person-attr/model-proc.json device=CPU ! \
    gvametaconvert format=json ! gvawatermark ! videoconvert ! autovideosink sync=false
```

The detection model is exported from YOLOv8, the classify model is from this repo: https://github.com/hungtrieu07/pedestrian-attribute-recognition-pytorch

I'm using this code snippet for get the result after `gvainference` element:

```
from gstgva import VideoFrame

def process_frame(frame: VideoFrame):
    for roi in frame.regions():
        for tensor in roi.tensors():
            if tensor['attribute_name'] == 'person-attributes':
                print(f"Attributes Label: {tensor['label']}, Confidence: {tensor['confidence']}")
    return True
```

Terminal output:
```
Attributes Label: hairShort upperBodyBlack upperBodyGrey upperBodyLongSleeve lowerBodyBlack carryingMessengerBag personalLess45 personalFemale, Confidence: 0.9929405450820923
Attributes Label: hairShort upperBodyBlack upperBodyLongSleeve lowerBodyBlack carryingMessengerBag personalLess45 personalMale, Confidence: 0.9911113977432251
Attributes Label: hairShort upperBodyBlack upperBodyGrey upperBodyLongSleeve lowerBodyBlack personalLess45 personalFemale, Confidence: 0.9977957606315613
Attributes Label: hairShort upperBodyBlack upperBodyLongSleeve lowerBodyBlack personalLess45 personalMale, Confidence: 0.9703301191329956
Attributes Label: hairLong upperBodyBlack upperBodyLongSleeve lowerBodyBlack carryingMessengerBag personalLess45 personalFemale, Confidence: 0.9878377318382263
```

In the original person attribute recognition model, the output shape is [1, 45], each value is attribute confidence score, scale from 0 to 1; but in above terminal output, I don't know how they be calculated; and I need to process on each confidence score to improve the output on display.

**Comment by hungtrieu07** [comment]
> The shown pipeline doesn't contain a `gvainference`, do you mean after `gvadetect` and/or after `gvaclassif` you add another plugin `gvapython` using the shown code-snippet to print the attributes?
> 
> What does the model-proc file `ov_exported_model/person-attr/model-proc.json` look like?
> 
> Do you expect a confidence level for each of the attributes, but you see only one? (you expect each attribute to be printed on a separate line and each attribute has its own confidence level?)

Oh, I commented a wrong pipeline, this is the correct pipeline:
```
GST_DEBUG=3 gst-launch-1.0 filesrc location=../video_data/1740731094984.mp4 ! decodebin ! videoconvert ! \
    gvadetect model=./ov_exported_model/person_vehicle/FP16.xml device=CPU ! queue ! \
    gvaclassify model=./onnx_models/openvino_model/model.xml model-proc=./onnx_models/openvino_model/model-proc.json object-class="person" device=CPU ! queue ! \
    gvapython module=./process_attributes.py ! queue ! \
    gvawatermark ! videoconvert ! autovideosink sync=false
```

This is my `model-proc` file:
```
{
    "json_schema_version": "2.2.0",
    "input_preproc": [
        {
            "params": {
                "resize": "aspect-ratio",
                "range": [
                    0.0,
                    1.0
                ]
            }
        }
    ],
    "output_postproc": [
        {
            "layer_name": "output",
            "attribute_name": "person-attributes",
            
            "labels": [
                "accessoryHat",
                "",
                "hairLong",
                "",
                "hairShort",
                "",
                "upperBodyShortSleeve",
                "",
                "upperBodyBlack",
                "",
                "upperBodyBlue",
                "",
                "upperBodyBrown",
                "",
                "upperBodyGreen",
                "",
                "upperBodyGrey",
                "",
                "upperBodyOrange",
                "",
                "upperBodyPink",
                "",
                "upperBodyPurple",
                "",
                "upperBodyRed",
                "",
                "upperBodyWhite",
                "",
                "upperBodyYellow",
                "",
                "upperBodyLongSleeve",
                "",
                "lowerBodyShorts",
                "",
                "lowerBodyShortSkirt",
                "",
                "lowerBodyBlack",
                "",
                "lowerBodyBlue",
                "",
                "lowerBodyBrown",
                "",
                "lowerBodyGreen",
                "",
                "lowerBodyGrey",
                "",
                "lowerBodyOrange",
                "",
                "lowerBodyPink",
                "",
                "lowerBodyPurple",
                "",
                "lowerBodyRed",
                "",
                "lowerBodyWhite",
                "",
                "lowerBodyYellow",
                "",
                "lowerBodyLongSkirt",
                "",
                "footwearLeatherShoes",
                "",
                "footwearSandals",
                "",
                "footwearShoes",
                "",
                "footwearSneaker",
                "",
                "carryingBackpack",
                "",
                "carryingMessengerBag",
                "",
                "carryingLuggageCase",
                "",
                "carryingSuitcase",
                "",
                "personalLess30",
                "",
                "personalLess45",
                "",
                "personalLess60",
                "",
                "personalLarger60",
                "",
                "personalLess15",
                "",
                "personalMale",
                "",
                "personalFemale",
                ""
            ],
            "converter": "label",
            "method": "compound"
        }
    ]
}
```

I need the confidence score on each attribute of the model. You can get the test video file and model from this drive link: 

https://drive.google.com/drive/folders/1k3f7xp28WFhI0JFArXTgJNsyOyezijtq?usp=sharing

**Comment by hungtrieu07** [comment]
> In the model-proc file you use the method `"method": "compound"`.
> 
> From "
> 
> [dlstreamer/src/monolithic/gst/inference_elements/common/post_processor/converters/to_tensor/label.h](https://github.com/dlstreamer/dlstreamer/blob/b97196d6129a7481998839b3429b9846d849d636/src/monolithic/gst/inference_elements/common/post_processor/converters/to_tensor/label.h#L24)
> 
> Line 24 in [b97196d](/dlstreamer/dlstreamer/commit/b97196d6129a7481998839b3429b9846d849d636)
> 
>  enum class Method { Max, SoftMax, Compound, Multi, SoftMaxMulti, Index, Default = Max }; 
> " there are several methods available:
> `enum class Method { Max, SoftMax, Compound, Multi, SoftMaxMulti, Index, Default = Max };`
> 
> Have a look under
> 
> * [dlstreamer/src/cpu/tensor_postproc/tensor_postproc_label.cpp](https://github.com/dlstreamer/dlstreamer/blob/b97196d6129a7481998839b3429b9846d849d636/src/cpu/tensor_postproc/tensor_postproc_label.cpp#L158)
>       
>       
>            Line 158
>         in
>         [b97196d](/dlstreamer/dlstreamer/commit/b97196d6129a7481998839b3429b9846d849d636)
>       
>     
>     
>       
>   
>           
>             
>              void run_compound(const float *data, size_t size, std::string &label, double &confidence) const {
> * [dlstreamer/src/monolithic/gst/inference_elements/common/post_processor/converters/to_tensor/label.cpp](https://github.com/dlstreamer/dlstreamer/blob/b97196d6129a7481998839b3429b9846d849d636/src/monolithic/gst/inference_elements/common/post_processor/converters/to_tensor/label.cpp#L61)
>       
>       
>            Line 61
>         in
>         [b97196d](/dlstreamer/dlstreamer/commit/b97196d6129a7481998839b3429b9846d849d636)
>       
>     
>     
>       
>   
>           
>             
>              void compound_method(const DataType *data, size_t size, const std::vector<std::string> &labels, double threshold,
> 
> to see how the `compound` is done - and resulting in one confidence value only.
> 
> From "https://dlstreamer.github.io/dev_guide/model_proc_file.html" it looks like for `gvaclassify` you could specify a confidence-threshold (or default value 0.5 is used).
> 
> From e.g. "https://community.intel.com/t5/Intel-Distribution-of-OpenVINO/Result-on-display-when-using-gvaclassify-element-on-custom/m-p/1678249" the `compound` method was mentioned to "merge" the different attributes into one "label"... but then only one confidence value remains.
> 
> Maybe someone from the DLStreamer dev-team or customer-enabling team can comment.

I will open a new thread on Intel Community about this problem. Thank you 😄 Closing this issue.
